### PR TITLE
Test infrastructure for vmstat & memstat.

### PR DESCRIPTION
**Pull request by anjalidk**

### CODE REVIEW

File: tests/MEM/mem_test.c [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <resource.h>
+
+int main(int argc, char **argv)
+{
+        struct memstat mem, exist;
```
Comment (hallyn) [inline tests/MEM/mem_test.c:27]: This has spaces instead of a tab.

File: tests/MEM/mem.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export LD_LIBRARY_PATH=../..
```
Comment (hallyn) [inline tests/MEM/mem.sh:1]: This and vm.sh

1. aren't called anywhere, is that happening in another pr that will hook up .github/workflows to do CI?
2. aren't doing set -e or a #!/bin/sh

File: tests/MEM/mem_test.c [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	unsigned long value;
+	FILE *fp;
+
+	res_read(RES_MEM_INFOALL, &mem, sizeof(mem), NULL, 0, 0);
+	res_exist(RES_MEM_INFOALL, &exist, sizeof(exist), NULL, 0, 0);
```
Comment (hallyn) [inline tests/MEM/mem_test.c:32]: You're ignoring the return values of both res_read and res_exist.

File: resource.c [hunk 4]
```diff
@@ -209,6 +209,15 @@ int res_read(int res_id, void *out, size_t out_sz, void **hint, int pid, int fla
 	return 0;
 }
 
+int res_exist(int res_id, void *out, size_t out_sz, void *hint, int pid, int flags)
+{
+	if (res_id >= VM_MIN && res_id < VM_MAX)
+		return getvmexist(res_id, out, out_sz, hint, flags);
```
Comment (hallyn) [inline resource.c:215]: getvmexist() doesn't seem to exist any more.  It is in resvm.h, but was removed

```
commit b408e61e7ba256c20af2ce7cd811bf79529f8b63
Author: Anjali Kulkarni <anjali.k.kulkarni@oracle.com>
Date:   Fri Apr 21 08:58:38 2023 -0700
```

File: tests/MEM/mem_test.c [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	fp = fopen ("./mem_info.txt", "w");
```
Comment (hallyn) [inline tests/MEM/mem_test.c:33]: not checking fopen() return value.

**Review by hallyn — CHANGES_REQUESTED** [review]

**Comment by anjalidk** [comment]
From: Serge Hallyn ***@***.***>
Date: Saturday, July 8, 2023 at 11:14 AM
To: lxc/libresource ***@***.***>
Cc: Anjali Kulkarni ***@***.***>, Author ***@***.***>
Subject: Re: [lxc/libresource] Test infrastructure for vmstat & memstat. (PR #23)

@hallyn requested changes on this pull request.

________________________________

In tests/MEM/mem_test.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/23*discussion_r1257323657__;Iw!!ACWV5N9M2RV99hQ!JV5VntV3dJ2OZI-YaVmKBwHT-9fNBTKgBRnutPa7XxeetTOTaeSuqyC_c0BEK9K0wr0cdsAD1_EfKZCLawV-Rzqco_ko1RQ$>:

> + * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the

+ * GNU Lesser General Public License for more details.

+ *

+ * You should have received a copy of the GNU Lesser General Public License

+ * along with libresource. If not, see <http://www.gnu.org/licenses/><https://urldefense.com/v3/__http:/www.gnu.org/licenses/*3E__;JQ!!ACWV5N9M2RV99hQ!JV5VntV3dJ2OZI-YaVmKBwHT-9fNBTKgBRnutPa7XxeetTOTaeSuqyC_c0BEK9K0wr0cdsAD1_EfKZCLawV-Rzqcwo7qBEU$>.

+ */

+

+#include <stdio.h>

+#include <stdlib.h>

+#include <string.h>

+#include <unistd.h>

+#include <resource.h>

+

+int main(int argc, char **argv)

+{

+        struct memstat mem, exist;

This has spaces instead of a tab.

ANJALI> Will change.

________________________________

In tests/MEM/mem.sh<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/23*discussion_r1257323795__;Iw!!ACWV5N9M2RV99hQ!JV5VntV3dJ2OZI-YaVmKBwHT-9fNBTKgBRnutPa7XxeetTOTaeSuqyC_c0BEK9K0wr0cdsAD1_EfKZCLawV-RzqcEZIGSxY$>:

> @@ -0,0 +1,8 @@

+export LD_LIBRARY_PATH=../..

This and vm.sh

  1.  aren't called anywhere, is that happening in another pr that will hook up .github/workflows to do CI?
  2.  aren't doing set -e or a #!/bin/sh
ANJALI> Yes, there is another person who will work on testing this and adding QA infrastructure, I will ask her to look into this via another PR.
________________________________

In tests/MEM/mem_test.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/23*discussion_r1257324204__;Iw!!ACWV5N9M2RV99hQ!JV5VntV3dJ2OZI-YaVmKBwHT-9fNBTKgBRnutPa7XxeetTOTaeSuqyC_c0BEK9K0wr0cdsAD1_EfKZCLawV-RzqcXC57FH8$>:

> + */

+

+#include <stdio.h>

+#include <stdlib.h>

+#include <string.h>

+#include <unistd.h>

+#include <resource.h>

+

+int main(int argc, char **argv)

+{

+        struct memstat mem, exist;

+       unsigned long value;

+       FILE *fp;

+

+       res_read(RES_MEM_INFOALL, &mem, sizeof(mem), NULL, 0, 0);

+       res_exist(RES_MEM_INFOALL, &exist, sizeof(exist), NULL, 0, 0);

You're ignoring the return values of both res_read and res_exist.

ANJALI> Will check.

________________________________

In resource.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/23*discussion_r1257324284__;Iw!!ACWV5N9M2RV99hQ!JV5VntV3dJ2OZI-YaVmKBwHT-9fNBTKgBRnutPa7XxeetTOTaeSuqyC_c0BEK9K0wr0cdsAD1_EfKZCLawV-RzqcqW4wtt0$>:

> @@ -209,6 +209,15 @@ int res_read(int res_id, void *out, size_t out_sz, void **hint, int pid, int fla

        return 0;

 }



+int res_exist(int res_id, void *out, size_t out_sz, void *hint, int pid, int flags)

+{

+       if (res_id >= VM_MIN && res_id < VM_MAX)

+              return getvmexist(res_id, out, out_sz, hint, flags);

getvmexist() doesn't seem to exist any more. It is in resvm.h, but was removed

commit b408e61e7ba256c20af2ce7cd811bf79529f8b63

Author: Anjali Kulkarni ***@***.***>

Date:   Fri Apr 21 08:58:38 2023 -0700

ANJALI> Will fix this.

________________________________

In tests/MEM/mem_test.c<https://urldefense.com/v3/__https:/github.com/lxc/libresource/pull/23*discussion_r1257324349__;Iw!!ACWV5N9M2RV99hQ!JV5VntV3dJ2OZI-YaVmKBwHT-9fNBTKgBRnutPa7XxeetTOTaeSuqyC_c0BEK9K0wr0cdsAD1_EfKZCLawV-RzqcE5Gqu_o$>:

> +

+#include <stdio.h>

+#include <stdlib.h>

+#include <string.h>

+#include <unistd.h>

+#include <resource.h>

+

+int main(int argc, char **argv)

+{

+        struct memstat mem, exist;

+       unsigned long value;

+       FILE *fp;

+

+       res_read(RES_MEM_INFOALL, &mem, sizeof(mem), NULL, 0, 0);

+       res_exist(RES_MEM_INFOALL, &exist, sizeof(exist), NULL, 0, 0);

+       fp = fopen ("./mem_info.txt", "w");

not checking fopen() return value.

ANJALI> Will check.

**Review by hallyn — APPROVED** [review]

**Comment by anjalidk** [comment]
Hi Serge,
Are these changes merged into master? I do not see them when I checkout a fresh libresource..
Anjali

**Comment by hallyn** [comment]
> Hi Serge, Are these changes merged into master? I do not see them when I checkout a fresh libresource.

No, I had not yet merged, as I was waiting for you to rebase, squash and push.  But I'll just let github do the squash-and-merge.

**Pull request merged**